lib/Sema/CSApply.cpp

```diff
@@ -2274,91 +2274,97 @@ namespace {
-      assert(interpolationProto && "Missing string interpolation protocol?");
+      auto loc = expr->getStartLoc();
 
-      DeclName name(tc.Context, DeclBaseName::createConstructor(),
-                    { tc.Context.Id_stringInterpolation });
-      auto member
-        = findNamedWitnessImpl(
-            tc, dc, type,
-            interpolationProto, name,
-            diag::interpolation_broken_proto);
-
-      DeclName segmentName(tc.Context, DeclBaseName::createConstructor(),
-                           { tc.Context.Id_stringInterpolationSegment });
-      auto segmentMember
-        = findNamedWitnessImpl(
-            tc, dc, type, interpolationProto, segmentName,
-            diag::interpolation_broken_proto);
-      if (!member ||
-          !segmentMember ||
-          !isa<ConstructorDecl>(member.getDecl()) ||
-          !isa<ConstructorDecl>(segmentMember.getDecl()))
-        return nullptr;
+      auto buildProtocolInitCall =
+        [&](KnownProtocolKind protocolKind, Type type, 
+            ArrayRef<Identifier> argLabels, ArrayRef<Expr *> args) -> Expr * {
 
-      // Build a reference to the init(stringInterpolation:) initializer.
-      // FIXME: This location info is bogus.
-      auto *typeRef = TypeExpr::createImplicitHack(expr->getStartLoc(), type,
-                                                   tc.Context);
+        auto proto = tc.getProtocol(loc, protocolKind);
+        assert(proto && "Missing string interpolation protocol?");
 
-      Expr *memberRef =
-        new (tc.Context) MemberRefExpr(typeRef,
-                                       expr->getStartLoc(),
-                                       member.getDecl(),
-                                       DeclNameLoc(expr->getStartLoc()),
-                                       /*Implicit=*/true);
-      cs.cacheSubExprTypes(memberRef);
-      cs.setSubExprTypes(memberRef);
-      bool failed = tc.typeCheckExpressionShallow(memberRef, cs.DC);
-      cs.cacheExprTypes(memberRef);
-      assert(!failed && "Could not reference string interpolation witness");
-      (void)failed;
+        auto conformance =
+          tc.conformsToProtocol(type, proto, cs.DC,
+                              ConformanceCheckFlags::InExpression);
+        assert(conformance && "string interpolation type conforms to protocol");
 
-      // Create a tuple containing all of the segments.
-      SmallVector<Expr *, 4> segments;
-      SmallVector<Identifier, 4> names;
-      ConstraintLocatorBuilder locatorBuilder(cs.getConstraintLocator(expr));
-      auto getType = [&](const Expr *E) -> Type {
-        return cs.getType(E);
+        DeclName constrName(tc.Context, DeclBaseName::createConstructor(), argLabels);
+        
+        Expr *base = TypeExpr::createImplicitHack(loc, type,
+                                                  tc.Context);
+        Expr *semanticExpr = tc.callWitness(base, dc, proto, *conformance,
```

Comment:
Does this still rely on overload resolution or is it always calling the actual witness for the requirement?

Reply:
This code (which generates the `init(literalCapacity:interpolationCount:)` and `init(stringInterpolation:)` calls) ought to always call the actual witness. That is what `TypeChecker::callWitness()` does, right? It does solve a constraint system, but I think it only uses it to calculate argument conversions, not to look for overloads.

Calls to `appendInterpolation` methods do use overload resolution; this is intended and necessary for certain desirable features and optimizations. Calls to the `appendLiteral(_:)` method also use overload resolution; this isn't really intended, but it'd be difficult to treat them differently from `appendInterpolation(_:)` without a more specialized AST representation.

Reply:
As long as the overloaded call is built at parse time or CSGen and resolved by the time we get to CSApply it’s fine.

callWitness() is overkill for what it does — we don’t need to create a constraint system to look up a single member and build a call. But as long as you’re not relying on it for overload resolution I’ll be able to clean it up later. My plan is to get rid of callWitness() and any other places in CSApply where we solve new constraint systems. I think typeCheckExpressionShallow() was the other main case but you removed that which is great.